### PR TITLE
Fix training

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

- [x] Updates the `NeuronTrainer` to match Transformers version
- [x] Fixes the way mixed-precision training is handled. It was breaking multiple trainings such as Llama. We do not use `XLA_USE_BF16` and `XLA_DOWNCAST_BF16` flags anymore following instructions from [here](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/appnotes/torch-neuronx/migration-from-xla-downcast-bf16.html)
- [x] Fixes support for gradient clipping: now it always happens between the reduction of the gradients across the devices and the optimizer step. The gradient norm is also now always reported when logging.
- [x] Adds support for `ignore_index` in the `parallel_cross_entropy_loss`. There was a big issue in training when using TP, the model was not learning. After investigating, it was linked to the input being padded and the vanilla `parallel_cross_entropy` from `neuronx_distributed` not supporting `ignore_index`:
     - First, `loss.mean()` does not work in this case because the loss for the ignored tokens is not zeroed. 
     - Second, the ignored tokens contributed to the gradient, which effectively destroys training.

For now DP + TP can lead to compilation issues with SDK 2.20, but they seem to be gone with SDK 2.21.

## Tests performed

- [x] Llama (`HuggingFaceTB/SmolLM2-135M-Instruct`) can overfit with dp=1 tp=1
- [x] Llama (`HuggingFaceTB/SmolLM2-135M-Instruct`) + LoRA can overfit with dp=1 tp=1
- [x] Llama (`meta-llama/Llama-3.2-1B`) can overfit with dp=1 tp=2
- [x]  Llama + LoRA (`meta-llama/Llama-3.2-1B`) can overfit with dp=1 tp=2
- [x] Llama (`meta-llama/Llama-3.2-1B`) can overfit with dp=4 tp=2 **(Only tested on SDK 2.21, otherwise compiler error)**
- [x] Actual training of Llama (`meta-llama/Llama-3.2-1B`) dp=4 tp=2 **on SDK 2.21** and compared to GPUs

W B Chart 20_02_2025 17_35_18

W B Chart 20_02_2025 17_35_38

## To be done in following PRs

- Add `AdamW_FP32Params` for a more stable training in mixed-precision

### CODE REVIEW

File: optimum/neuron/accelerate/accelerator.py [hunk 1]
```diff
@@ -224,8 +223,8 @@ def prepare_data_loader(self, data_loader: DataLoader, device_placement: Optiona
                 data_loader, num_replicas=num_replicas, rank=rank, force_drop_last=force_drop_last
             )
             # No need to wrap the dataloader if we are using pipeline parallelism.
-            if self.state.mp_plugin.pipeline_parallel_size == 1:
-                data_loader = MpDeviceLoader(data_loader, self.device)
+            # if self.state.mp_plugin.pipeline_parallel_size == 1:
+            #     data_loader = MpDeviceLoader(data_loader, self.device)
```
Comment (tengomucho) [inline optimum/neuron/accelerate/accelerator.py:227]: remove this code instead of commenting it out

**Review by tengomucho — COMMENTED** [review]
I find difficult to understand what your changes have accomplished, can you give more details about that please?
Also, would you mind pointing to a test (or example) that now works after your changes?

File: optimum/neuron/trainers.py [hunk 2]
```diff
@@ -189,6 +188,9 @@ def __init__(self, *args, **kwargs):
 
         super().__init__(*args, **kwargs)
 
+        # This is important to avoid changing the way the loss is computed.
+        # self.model_accepts_loss_kwargs = False
+
```
Comment (dacorvo) [inline optimum/neuron/trainers.py:193]: You should remove this

**Review by dacorvo — COMMENTED** [review]
Several failures in the CI:
- style check,
- errors in the training code directly related to some of the changes (_PARALLEL_CROSS_ENTROPY_SHOULD_PRESERVE_INPUT is not found).

Maybe this should be rebased on the 2.21.1 branch once it is merged.

**Review by JingyaHuang — COMMENTED** [review]
Not too much to comment about, kudos for finding the fix!!

(will approve when the CIs pass)

**Comment by michaelbenayoun** [comment]
> Several failures in the CI:
> 
> * style check,
> * errors in the training code directly related to some of the changes (_PARALLEL_CROSS_ENTROPY_SHOULD_PRESERVE_INPUT is not found).
> 
> Maybe this should be rebased on the 2.21.1 branch once it is merged.

I am working on fixing the CI failures.
If you want you can merge the 2.21.1 branch, then I will rebase and adapt: some changes in the PR are linked to 2.20, so if we are going to move to 2.21 very soon it does not make sense to add them.

**Comment by dacorvo** [comment]
I bumped the SDK version to 2.21.1: you can now rebase your branch and drop the 2.20 specifics.

**Comment by tengomucho** [comment]
These changes seem to make the SFT tutorial converge again.

File: optimum/neuron/distributed/utils.py [hunk 3]
```diff
@@ -761,31 +771,48 @@ def maybe_load_linear_weight_to_gqa_qkv_column_parallel_linear(
             "A linear's layer WeightInformation or a linear layer to copy the weights from need to specified."
         )
 
-    proj_name = weight_name[-1]
-    weight = getattr(layer, weight_name)
-    bias = getattr(layer, f"bias_{proj_name}")
+    # proj_name = weight_name[-1]
```
Comment (dacorvo) [inline optimum/neuron/distributed/utils.py:774]: You can remove this commented line.

File: optimum/neuron/distributed/utils.py [hunk 4]
```diff
@@ -1749,3 +1783,161 @@ def __enter__(self):
...
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        # Retreive tensors from the forward path.
+        softmax, target_mask, masked_target_1d, is_non_ignore_index_mask = ctx.saved_tensors
+        label_smoothing, vocab_size = ctx.label_smoothing, ctx.vocab_size
+
+        reduction = ctx.reduction
+
+        # All the inputs have softmax as thier gradient.
```
Comment (dacorvo) [inline optimum/neuron/distributed/utils.py:1902]:
```suggestion
        # All the inputs have softmax as their gradient.
```

File: optimum/neuron/trainers.py [hunk 5]
```diff
@@ -438,11 +454,12 @@ def training_step(self, model: torch.nn.Module, inputs: Dict[str, Union[torch.Te
                 use_bf16 = self.accelerator.state.mixed_precision == "bf16"
                 dtype = torch.bfloat16 if use_bf16 else torch.float32
                 loss = torch.tensor(0, dtype=dtype).to(xm.xla_device())
-            else:
-                loss = loss.detach()
+
+            # TODO: handle with num_items_in_batch
```
Comment (dacorvo) [inline optimum/neuron/trainers.py:458]: Maybe raise an exception here if num_items_in_batch is not None ?

**Review by dacorvo — CHANGES_REQUESTED** [review]
A few nits and maybe be careful when deleting test repos, otherwise looks good to me. Thanks !

**Review by dacorvo — APPROVED** [review]
Thanks !

**Comment by SBrandeis** [comment]
GG
